fix: Fix subscriptions sending twice

webjunkie opened the pull request:
## Problem

Daily subscriptions are sent twice.

Theory: Since they run 5 minutes before the full hour, next delivery date will be scheduled again on same day (one hour later). Only then the `now` is larger then the intended date and next run will be scheduled correctly in a day.

Follow up from #21419

## Changes

Revert  #21419 but add `max()` so that we never take a date in the past.

## Does this work well for both Cloud and self-hosted?

it doesn't have an impact

## How did you test this code?

- Cover both this case and the original case in tests.

webjunkie committed:
Fix subscriptions sending twice

webjunkie commented on posthog/models/subscription.py:
```diff
@@ -130,6 +130,8 @@ def save(self, *args, **kwargs) -> None:
         # Only if the schedule has changed do we update the next delivery date
         if not self.id or str(self._rrule) != str(self.rrule):
             self.set_next_delivery_date()
+            if "update_fields" in kwargs:
+                kwargs["update_fields"].append("next_delivery_date")
```
This is now recommended to add in Django 4.2, so doing it here while on it.

https://docs.djangoproject.com/en/4.2/releases/4.2/#setting-update-fields-in-model-save-may-now-be-required

webjunkie committed:
Adjust calculation, add tests

webjunkie committed:
Add another test

webjunkie committed:
Merge branch 'master' into fix/subscriptions-twice

webjunkie committed:
Merge branch 'master' into fix/subscriptions-twice

mariusandra commented:
Seems like there's a legit test failure:

<img width="1012" alt="image" src="https://github.com/PostHog/posthog/assets/53387/862af4f1-8cfe-4489-8df2-2e9e54e70f83">

webjunkie committed:
Add buffer as well

webjunkie commented:
Yes, saw it after all 😅 
Buried after some other stuff... which I thought is CI trouble.

neilkakkar commented:
yeah a fix for this CI spew is here: https://github.com/PostHog/posthog/pull/21613 - getting there

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

webjunkie committed:
Adjust one other test

webjunkie committed:
Merge branch 'master' into fix/subscriptions-twice

webjunkie committed:
Merge branch 'master' into fix/subscriptions-twice

webjunkie committed:
Adjust one other test

If we create subscriptions in the buffer that we have before the full
hour we shouldn't expect them to run right in this hour. This is
getting wonky.

webjunkie committed:
Merge branch 'master' into fix/subscriptions-twice

webjunkie committed:
Merge branch 'master' into fix/subscriptions-twice
